relay-server/src/utils/metrics_rate_limits.rs

```diff
@@ -130,112 +131,229 @@ pub enum ExtractionMode {
     Duration,
 }
 
-/// Return value of [`count_metric_bucket`], containing the extracted
-/// count of transactions and wether they have associated profiles.
-#[derive(Debug, Clone, Copy)]
-struct TransactionCount {
-    /// Number of transactions.
-    pub count: usize,
-    /// Whether the transactions have associated profiles.
-    pub has_profile: bool,
+/// The return value of [`count_metric_bucket`].
+///
+/// Contains the count of total transactions or spans that went into this bucket.
+#[derive(Debug, Default, Clone)]
+pub enum BucketSummary {
+    Transactions {
+        count: usize,
+        has_profile: bool,
+    },
+    Spans(usize),
+    #[default]
+    None,
+}
+
+impl BucketSummary {
+    fn to_counts(&self) -> EntityCounts {
+        match *self {
+            BucketSummary::Transactions { count, has_profile } => EntityCounts {
+                transactions: Some(count),
+                spans: None,
+                profiles: if has_profile { count } else { 0 },
+            },
+            BucketSummary::Spans(count) => EntityCounts {
+                transactions: None,
+                spans: Some(count),
+                profiles: 0,
+            },
+            BucketSummary::None => EntityCounts::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct SummarizedBucket {
+    bucket: Bucket,
+    summary: BucketSummary,
+}
+
+/// Contains the total counts of limitable entities represented by a batch of metrics.
+#[derive(Debug, Default, Clone)]
+struct EntityCounts {
+    /// The number of transactions represented in the current batch.
+    ///
+    /// - `None` if the batch does not contain any transaction metrics.
+    /// - `Some(0)` if the batch contains transaction metrics, but no `usage` count.
+    /// - `Some(n > 0)` if the batch contains a `usage` count for transactions.
+    ///
+    /// The distinction between `None` and `Some(0)` is needed to decide whether or not a rate limit
+    /// must be checked.
+    transactions: Option<usize>,
+    /// The number of spans represented in the current batch.
+    ///
+    /// - `None` if the batch does not contain any transaction metrics.
+    /// - `Some(0)` if the batch contains transaction metrics, but no `usage` count.
+    /// - `Some(n > 0)` if the batch contains a `usage` count for spans.
+    ///
+    /// The distinction between `None` and `Some(0)` is needed to decide whether or not a rate limit
+    /// must be checked.
+    spans: Option<usize>,
+    /// The number of profiles represented in the current batch.
+    ///
+    /// We do not explicitly check the rate limiter for profiles, so there is no need to
+    /// distinguish between `None` and `Some(0)`.
+    profiles: usize,
```

Comment:
baby nit: Maybe have the profiles part of the transaction count `transactions: Option<TransactionCountWithProfiles>`?

Reply:
Yeah that would work